app/services/notifications/sync_system_notifications_service.rb

```diff
@@ -81,12 +81,26 @@ def save_new_notifications
             .where(source_id: attrs[:source_id]).first_or_initialize(attrs)
 
         if n.new_record?
-          n.users = User.all
-          n.save!
+          save_notification n
         elsif n.last_time_changed_at < attrs[:last_time_changed_at]
           n.update_attributes!(attrs)
         end
       end
     end
+
+    def save_notification(notification)
+      ActiveRecord::Base.transaction do
+        notification.save!
+
+        user_system_notifications = []
+        User.pluck(:id).each do |user_id|
```

Comment:
Does it load IDs of all users in the single array?

Reply:
True. For array with 100k users ObjectSpace.memsize_of(obj) returns 780 KB. And limit for items nr in an array should be around 500mio.
Do you see any issues?

Reply:
Use find_each

Reply:
@radiokills It is just an array of ID-s, not active record objects... But yeah... I will.

Reply:
User.in_batches(of: 10) do |user_ids| puts user_ids.pluck(:id) end

Reply:
@okriuchykhin do we care about size of `User.pluck(:id)` or `user_system_notifications` array?

Reply:
Default is 1000, it should be fine

Reply:
On 100k rows in `users`
```
User.all.pluck(:id)
   (35.2ms)  SELECT "users"."id" FROM "users"
```
```
User.in_batches(of: 10) do |user_ids| puts user_ids.pluck(:id) end
   takes: 20seconds
```
```
User.in_batches(of: 10000) do |user_ids| puts user_ids.pluck(:id) end
   10 times around 8ms
```

Reply:
Ok... but what is the idea? To make inside of `in_batches` loop `SystemNotification.import` to prevent to have too big `user_system_notifications[]` or just to have a smaller array of `user_ids`?

Reply:
It is not about the speed, but to keep resource allocation under control

Reply:
To make inside of in_batches loop SystemNotification.import to prevent to have too big user_system_notifications[], user_ids[] will be same size

Reply:
As example:

User.in_batches(of:10) do |user_ids| user_system_notifications = user_ids.pluck(:id).collect { |item| Hash[:user_id, item, :system_notification_id, 33] }; pp user_system_notifications end

Reply:
Better version:

User.find_in_batches(batch_size: 10) do |user_ids| user_system_notifications = user_ids.pluck(:id).collect { |item| Hash[:user_id, item, :system_notification_id, 33] }; pp user_system_notifications end